fvtest/tril/tril/parser.cpp

```diff
@@ -0,0 +1,482 @@
+/*******************************************************************************
+ * Copyright (c) 2000, 2019 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "ast.h"
+#include "ast.hpp"
+#include <stdio.h>
+#include <string>
+#include <iostream>
+#include <vector>
+#include <iterator>
+
+
+static ASTNode* trees;
+
+int MAX_NUM = 128;
+
+class Token {
+  public:
+    enum Type {
+        ID = 0,
+        STRING_,
+        STRING_0,
+        INT,
+        WHITESPACE,
+        AT,
+        ZERO,
+        HEXINT,
+        QUOT,
+        LPAREN,
+        RPAREN,
+        LPAREN_SQ,
+        RPAREN_SQ,
+        BECOMES,
+        COMMA,
+        COMMENT,
+        DOUBLE_
+    };
+
+    private:
+        Type type;
+        std::string value;
+    
+
+    public:
+        Token(Type type, std::string value):
+	        type(type), value(value) {}
+        Type getType() const { return type; }
+        void setType(Token::Type t) {type = t ;}
+        const std::string &getValue() const {return value ;}
+};
+
+class TokenFailure {
+	std::string message;
+
+public:
+	TokenFailure(std::string message);
+	const std::string &what() const;
+};
+
+TokenFailure::TokenFailure(std::string message) :
+	message(message) {}
+
+const std::string &TokenFailure::what() const { return message; }
+
+int isNotNewLine(int c) { return c != '\n'; }
+
+int isNotQuot(int c) { return c != '"'; }
+
+int isHex(int c) {
+    if (c == 'a' || c == 'b' || c == 'c' || c == 'd' || c == 'e' || c =='f' ||
+        c == 'A' || c == 'B' || c == 'C' || c == 'D' || c == 'E' || c == 'F') {
```

Comment:
I too do not understand why digits are not included here. If we are able to include them we should make use of the standard library `std::isxdigit` [1] instead of rolling out our own.

[1] https://en.cppreference.com/w/cpp/string/byte/isxdigit